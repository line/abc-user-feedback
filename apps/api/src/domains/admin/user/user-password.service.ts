/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import * as bcrypt from 'bcrypt';
import { Repository } from 'typeorm';
import { Transactional } from 'typeorm-transactional';

import { CodeTypeEnum } from '@/shared/code/code-type.enum';
import { CodeService } from '@/shared/code/code.service';
import { ResetPasswordMailingService } from '@/shared/mailing/reset-password-mailing.service';
import type { ResetPasswordDto } from './dtos';
import { ChangePasswordDto } from './dtos';
import { UserEntity } from './entities/user.entity';
import { InvalidPasswordException, UserNotFoundException } from './exceptions';

@Injectable()
export class UserPasswordService {
  constructor(
    @InjectRepository(UserEntity)
    private readonly userRepo: Repository<UserEntity>,
    private readonly resetPasswordMailingService: ResetPasswordMailingService,
    private readonly codeService: CodeService,
  ) {}

  @Transactional()
  async sendResetPasswordMail(email: string) {
    const user = await this.userRepo.findOneBy({ email });
    if (!user) throw new UserNotFoundException();

    const code = await this.codeService.setCode({
      type: CodeTypeEnum.RESET_PASSWORD,
      key: email,
    });

    await this.resetPasswordMailingService.send({ email, code });
  }

  async resetPassword({ email, code, password }: ResetPasswordDto) {
    const user = await this.userRepo.findOneBy({ email });
    if (!user) throw new UserNotFoundException();

    const { error } = await this.codeService.verifyCode({
      type: CodeTypeEnum.RESET_PASSWORD,
      key: email,
      code,
    });

    if (error) throw error;

    return await this.userRepo.save(
      Object.assign(user, {
        hashPassword: await this.createHashPassword(password),
      }),
    );
  }

  @Transactional()
  async changePassword(dto: ChangePasswordDto) {
    const { newPassword, password, userId } = dto;
    const user = await this.userRepo.findOneBy({
      id: userId,
    });
    const originHashPassword = user.hashPassword;

    if (!bcrypt.compareSync(password, originHashPassword)) {
      throw new InvalidPasswordException();
    }

    return await this.userRepo.save(
      Object.assign(user, {
        hashPassword: await this.createHashPassword(newPassword),
      }),
    );
  }

  async createHashPassword(password: string) {
    const salt = await bcrypt.genSalt(10);
    const hashPassword = await bcrypt.hash(password, salt);
    return hashPassword;
  }
}
