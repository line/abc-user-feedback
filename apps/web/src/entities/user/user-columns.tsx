/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import { createColumnHelper } from '@tanstack/react-table';
import dayjs from 'dayjs';

import { Badge } from '@ufb/ui';

import { DATE_TIME_FORMAT, displayString, TableCheckbox } from '@/shared';

import UpdateUserPopover from './ui/update-user-popover.ui';
import type { UserMember } from './user.type';

const columnHelper = createColumnHelper<UserMember>();

export const getUserColumns = () => [
  columnHelper.display({
    id: 'select',
    header: ({ table }) => (
      <TableCheckbox
        checked={table.getIsAllRowsSelected()}
        indeterminate={table.getIsSomeRowsSelected()}
        onChange={table.getToggleAllRowsSelectedHandler()}
      />
    ),
    cell: ({ row }) => (
      <TableCheckbox
        checked={row.getIsSelected()}
        disabled={!row.getCanSelect()}
        indeterminate={row.getIsSomeSelected()}
        onChange={row.getToggleSelectedHandler()}
      />
    ),
    size: 50,
    enableSorting: false,
  }),
  columnHelper.accessor('email', {
    header: 'Email',
    enableSorting: false,
    size: 220,
  }),
  columnHelper.accessor('type', {
    header: 'Type',
    enableSorting: false,
    size: 80,
  }),
  columnHelper.accessor('name', {
    header: 'Name',
    enableSorting: false,
    cell: ({ getValue }) => displayString(getValue()),
    size: 120,
  }),
  columnHelper.accessor('department', {
    header: 'Department',
    cell: ({ getValue }) => displayString(getValue()),
    enableSorting: false,
    size: 120,
  }),
  columnHelper.accessor('members', {
    header: 'Project',
    cell: ({ getValue }) =>
      getValue().length > 0 ?
        <div className="flex flex-wrap gap-2">
          {getValue().map((member) => (
            <Badge key={member.id} type="secondary">
              {member.role.project.name}
            </Badge>
          ))}
        </div>
      : '-',
    enableSorting: false,
  }),
  columnHelper.accessor('createdAt', {
    header: 'Created',
    cell: ({ getValue }) => dayjs(getValue()).format(DATE_TIME_FORMAT),
    enableSorting: true,
  }),
  columnHelper.display({
    id: 'edit',
    header: 'Edit',
    cell: ({ row }) => <UpdateUserPopover user={row.original} />,
    size: 60,
  }),
];
