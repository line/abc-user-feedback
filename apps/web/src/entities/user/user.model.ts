/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import router from 'next/router';
import axios from 'axios';
import dayjs from 'dayjs';
import type { JwtPayload } from 'jwt-decode';
import { jwtDecode } from 'jwt-decode';
import { create } from 'zustand';

import { client, Path, sessionStorage } from '@/shared';
import type { Jwt } from '@/shared/types';

import type { User } from './user.type';

type State = { user: User | null };

type Action = {
  signInWithEmail: (input: {
    email: string;
    password: string;
  }) => Promise<void>;
  signInWithOAuth: (input: {
    code: string;
    callback_url?: string;
  }) => Promise<void>;
  signOut: () => Promise<void>;
  setUser: () => void;
  _signIn: (jwt: Jwt) => void;
};

export const useUserStore = create<State & Action>((set, get) => ({
  user: null,
  signInWithEmail: async ({ email, password }) => {
    const { data: jwt } = await axios.post('/api/login', { email, password });
    get()._signIn(jwt);
  },
  signInWithOAuth: async ({ code, callback_url }) => {
    const { data: jwt } = await axios.post('/api/oauth', {
      code,
      callback_url,
    });
    get()._signIn(jwt);
  },
  signOut: async () => {
    await axios.get('/api/logout');
    sessionStorage.removeItem('jwt');
    router.reload();
  },
  setUser: async () => {
    const jwt = sessionStorage.getItem('jwt');
    if (!jwt) return;

    const { sub, exp } = jwtDecode<JwtPayload>(jwt.accessToken);
    if (!sub || dayjs().isBefore(dayjs(exp))) get().signOut();
    else {
      const { data } = await client.get({
        path: '/api/admin/users/{id}',
        pathParams: { id: parseInt(sub) },
        options: { headers: { Authorization: `Bearer ${jwt.accessToken}` } },
      });
      set({ user: data });
    }
  },

  async _signIn(jwt) {
    sessionStorage.setItem('jwt', jwt);
    get().setUser();
    if (router.query.callback_url) {
      router.push(router.query.callback_url as string);
    } else {
      router.push({ pathname: Path.MAIN });
    }
  },
}));
