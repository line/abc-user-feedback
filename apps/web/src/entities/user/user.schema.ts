/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import { z } from 'zod';

export const userSchema = z.object({
  id: z.number(),
  email: z.string().email(),
  type: z.union([z.literal('SUPER'), z.literal('GENERAL')]),
  name: z.string().nullable(),
  department: z.string().nullable(),
  signUpMethod: z.union([z.literal('OAUTH'), z.literal('EMAIL')]),
});

export const memberSchema = z.object({
  id: z.number(),
  role: z.object({
    name: z.string(),
    project: z.object({
      id: z.number(),
      name: z.string(),
    }),
  }),
});

export const userMemberSchema = userSchema.merge(
  z.object({
    members: z.array(memberSchema),
    createdAt: z.string(),
  }),
);

export const updateUserSchema = userSchema.omit({
  id: true,
  signUpMethod: true,
});
