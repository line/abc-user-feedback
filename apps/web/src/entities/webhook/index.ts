export * from './ui';
export * from './webhook.type';
