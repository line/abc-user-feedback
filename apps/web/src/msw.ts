/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { http, HttpResponse } from 'msw';
import { setupServer } from 'msw/node';

import { signInWithOAuthMockHandlers } from './features/auth/sign-in-with-oauth/__mocks__/sign-in-with-oauth.mock-handler';
import type {
  OAIMethodPathKeys,
  OAIMethods,
  OAIPathParameters,
} from './shared';
import { convertToColonPath, getRequestUrl } from './shared';

export const server = setupServer(...signInWithOAuthMockHandlers);

export const simpleMockHttp = <
  TMethod extends OAIMethods,
  TPath extends OAIMethodPathKeys<TMethod>,
>({
  method,
  path,
  status = 200,
  params,
  data = {},
}: {
  method: TMethod;
  path: TPath;
  status?: 200 | 201 | 204 | 400 | 401 | 403 | 404 | 500;
  params?: OAIPathParameters<TPath, TMethod>;
  data?: Record<string, unknown>;
}) =>
  server.use(
    http[method](`${convertToColonPath(getRequestUrl(path, params))}`, () =>
      HttpResponse.json(data, { status }),
    ),
  );
