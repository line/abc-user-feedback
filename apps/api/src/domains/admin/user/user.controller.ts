/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UnauthorizedException,
  UseGuards,
} from '@nestjs/common';
import {
  ApiBearerAuth,
  ApiBody,
  ApiOkResponse,
  ApiTags,
} from '@nestjs/swagger';

import { PaginationRequestDto } from '@/common/dtos';
import { JwtAuthGuard } from '../auth/guards';
import { CurrentUser } from './decorators';
import { UserDto } from './dtos';
import {
  ChangePasswordRequestDto,
  DeleteUsersRequestDto,
  GetAllUsersRequestDto,
  ResetPasswordMailingRequestDto,
  ResetPasswordRequestDto,
  UpdateUserRequestDto,
  UserInvitationRequestDto,
} from './dtos/requests';
import { GetRolesByIdResponseDto } from './dtos/responses';
import { GetAllUserResponseDto } from './dtos/responses/get-all-user-response.dto';
import { UserTypeEnum } from './entities/enums';
import { SuperUser } from './super-user.decorator';
import { UserPasswordService } from './user-password.service';
import { UserService } from './user.service';

@ApiTags('users')
@Controller('/admin/users')
@ApiBearerAuth()
export class UserController {
  constructor(
    private readonly userService: UserService,
    private readonly userPasswordService: UserPasswordService,
  ) {}

  @ApiOkResponse({ type: GetAllUserResponseDto })
  @Get()
  @UseGuards(JwtAuthGuard)
  async getAllUsers(@Query() query: PaginationRequestDto) {
    const { limit, page } = query;
    const users = await this.userService.findAll({
      options: { limit, page },
    });
    return GetAllUserResponseDto.transform(users);
  }

  @ApiOkResponse({ type: GetAllUserResponseDto })
  @Post('/search')
  @HttpCode(200)
  @UseGuards(JwtAuthGuard)
  async searchUsers(@Body() body: GetAllUsersRequestDto) {
    const { limit, page, query, order } = body;
    const users = await this.userService.findAll({
      options: { limit, page },
      query,
      order,
    });
    return GetAllUserResponseDto.transform(users);
  }

  @Delete()
  @SuperUser()
  async deleteUsers(@Body() body: DeleteUsersRequestDto) {
    await this.userService.deleteUsers(body.ids);
  }

  @ApiOkResponse({ type: UserDto })
  @Get(':id')
  @UseGuards(JwtAuthGuard)
  async getUser(
    @Param('id', ParseIntPipe) id: number,
    @CurrentUser() user: UserDto,
  ) {
    if (id !== user?.id) throw new UnauthorizedException('');
    return UserDto.transform(await this.userService.findById(id));
  }

  @ApiOkResponse({ type: GetRolesByIdResponseDto })
  @Get(':userId/roles')
  @UseGuards(JwtAuthGuard)
  async getRoles(@Param('userId', ParseIntPipe) userId: number) {
    const roles = await this.userService.findRolesById(userId);
    return { roles };
  }

  @Delete(':id')
  @UseGuards(JwtAuthGuard)
  async deleteUser(
    @Param('id', ParseIntPipe) id: number,
    @CurrentUser() user: UserDto,
  ) {
    if (user.id !== id) throw new UnauthorizedException();
    await this.userService.deleteById(id);
  }

  @Put('/:id')
  @HttpCode(204)
  @UseGuards(JwtAuthGuard)
  async updateUser(
    @Param('id', ParseIntPipe) userId: number,
    @Body() dto: UpdateUserRequestDto,
    @CurrentUser() user: UserDto,
  ) {
    if (user.type === UserTypeEnum.GENERAL) {
      if (dto.type)
        throw new UnauthorizedException('GENERAL user cannot modify user type');
      if (user.id !== userId)
        throw new UnauthorizedException(
          'GENERAL user cannot modify other user',
        );
    }
    await this.userService.updateUser({ userId, ...dto });
  }

  @Post('invite')
  @SuperUser()
  async inviteUser(
    @Body() body: UserInvitationRequestDto,
    @CurrentUser() user: UserDto,
  ) {
    if (body.userType === UserTypeEnum.SUPER && body.roleId) {
      throw new BadRequestException('SUPER user must not have role');
    }
    await this.userService.sendInvitationCode({ ...body, invitedBy: user });
  }

  @ApiBody({ type: ResetPasswordMailingRequestDto })
  @Post('password/reset/code')
  async requestResetPassword(@Body('email') email: string) {
    await this.userPasswordService.sendResetPasswordMail(email);
  }

  @Post('password/reset')
  async resetPassword(@Body() body: ResetPasswordRequestDto) {
    await this.userPasswordService.resetPassword(body);
  }

  @Post('password/change')
  @UseGuards(JwtAuthGuard)
  async changePassword(
    @CurrentUser() user: UserDto,
    @Body() { newPassword, password }: ChangePasswordRequestDto,
  ) {
    await this.userPasswordService.changePassword({
      newPassword,
      password,
      userId: user.id,
    });
  }
}
