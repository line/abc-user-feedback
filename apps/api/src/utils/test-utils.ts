/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
import { faker } from '@faker-js/faker';
import { MailerModule } from '@nestjs-modules/mailer';
import { InjectionToken, Provider } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { TypeOrmModule } from '@nestjs/typeorm';
import { join } from 'path';
import { DataSource, Repository } from 'typeorm';
import { SnakeNamingStrategy } from 'typeorm-naming-strategies';
import {
  addTransactionalDataSource,
  initializeTransactionalContext,
} from 'typeorm-transactional';

import { jwtConfig, jwtConfigSchema } from '@/configs/jwt.config';
import { mySqlConfigSchema, mysqlConfig } from '@/configs/mysql.config';
import { smtpConfig, smtpConfigSchema } from '@/configs/smtp.config';
import { AuthService } from '@/domains/auth/auth.service';
import { UserDto } from '@/domains/user/dtos';
import { UserStateEnum } from '@/domains/user/entities/enums';
import { UserEntity } from '@/domains/user/entities/user.entity';
import { ConfigServiceType } from '@/types/config-service.type';

initializeTransactionalContext();

export const getMockProvider = (
  injectToken: InjectionToken,
  factory: unknown,
): Provider => ({ provide: injectToken, useFactory: () => factory });

export const TestConfig = ConfigModule.forRoot({
  load: [smtpConfig, jwtConfig, mysqlConfig],
  envFilePath: '.env.test',
  validate: (config) => ({
    ...smtpConfigSchema.validateSync(config),
    ...jwtConfigSchema.validateSync(config),
    ...mySqlConfigSchema.validateSync(config),
  }),
});

export const TestMailConfigModule = MailerModule.forRoot({
  transport: { url: '' },
});

export const TestTypeOrmConfig = TypeOrmModule.forRootAsync({
  imports: [ConfigModule],
  inject: [ConfigService],
  dataSourceFactory: async (options) => {
    const datasource = await new DataSource(options).initialize();
    return addTransactionalDataSource(datasource);
  },
  useFactory: (configService: ConfigService<ConfigServiceType>) => {
    const { main_url, sub_urls } = configService.get('mysql', {
      infer: true,
    });
    return {
      type: 'mysql',
      replication: {
        master: { url: main_url },
        slaves: sub_urls.map((url) => ({ url })),
      },
      entities: [join(__dirname, '../**/*.entity.{ts,js}')],
      logging: ['warn', 'error'],
      namingStrategy: new SnakeNamingStrategy(),
      timezone: '+00:00',
    };
  },
});

export const TestConfigs = [
  TestConfig,
  TestMailConfigModule,
  TestTypeOrmConfig,
];

export const getRandomEnumValue = <T>(anEnum: T): T[keyof T] => {
  const enumValues = Object.keys(anEnum) as Array<keyof T>;
  const randomIndex = faker.datatype.number(enumValues.length - 1);
  const randomEnumKey = enumValues[randomIndex];
  return anEnum[randomEnumKey];
};
export const getRandomEnumValues = <T>(anEnum: T): T[keyof T][] => {
  const enumValues = Object.values(anEnum) as Array<keyof T>;
  return faker.helpers.arrayElements(enumValues) as T[keyof T][];
};

export const clearEntities = async (repos: Repository<any>[]) => {
  for (const repo of repos) {
    await repo.query('set foreign_key_checks = 0');
    await repo.clear();
    await repo.query('set foreign_key_checks = 1');
  }
};

export const signInTestUser = async (
  dataSource: DataSource,
  authService: AuthService,
) => {
  const userRepo = dataSource.getRepository(UserEntity);
  const user = await userRepo.save({
    email: faker.internet.email(),
    state: UserStateEnum.Active,
    hashPassword: faker.internet.password(),
  });
  return { jwt: await authService.signIn(UserDto.transform(user)), user };
};

export const DEFAULT_FIELD_COUNT = 2;

export const createQueryBuilder: any = {
  setFindOptions: () =>
    jest.fn().mockImplementation(() => {
      return createQueryBuilder;
    }),
};

export const mockRepository = () => ({
  findOneBy: jest.fn(),
  findOne: jest.fn(),
  find: jest.fn(),
  findBy: jest.fn(),
  findAndCount: jest.fn(),
  findAndCountBy: jest.fn(),
  save: jest.fn(),
  create: jest.fn(),
  update: jest.fn(),
  upsert: jest.fn(),
  count: jest.fn(),
  remove: jest.fn(),
  createQueryBuilder: jest.fn(() => createQueryBuilder),
  query: jest.fn(),
});

export const MockOpensearchRepository = {
  createIndex: jest.fn(),
  deleteIndex: jest.fn(),
  putMappings: jest.fn(),
  createData: jest.fn(),
  getData: jest.fn(),
  updateData: jest.fn(),
  getTotal: jest.fn(),
};
