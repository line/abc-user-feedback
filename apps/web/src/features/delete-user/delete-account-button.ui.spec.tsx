/**
 * Copyright 2023 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

import { faker } from '@faker-js/faker';
import userEvent from '@testing-library/user-event';
import { http } from 'msw';

import * as user from '@/entities/user';

import { server, simpleMockHttp } from '@/msw';
import { render, screen, waitFor } from '@/test-utils';
import DeleteAccountButton from './delete-account-button.ui';

jest.mock('@/entities/user');

const TEST_USER: user.User = {
  id: faker.number.int(),
  email: faker.internet.email(),
  name: faker.string.alphanumeric(8),
  type: faker.helpers.arrayElement(['GENERAL', 'SUPER']),
  department: faker.helpers.arrayElement([null, faker.string.alphanumeric(8)]),
  signUpMethod: 'EMAIL',
};

server.use(
  http.get('/api/logout', () => {
    return;
  }),
);

describe('DeleteAccountButton', () => {
  test('match snapshot', () => {
    jest.spyOn(user, 'useUserStore').mockImplementation(() => ({
      signInWithEmail: jest.fn(),
      _signIn: jest.fn(),
      setUser: jest.fn(),
      signInWithOAuth: jest.fn(),
      signOut: jest.fn(),
    }));
    const component = render(<DeleteAccountButton user={TEST_USER} />);
    expect(component.container).toMatchSnapshot();
  });

  describe('Submittion', () => {
    const mockSignOut = jest.fn();
    beforeEach(async () => {
      jest.spyOn(user, 'useUserStore').mockImplementation(() => ({
        signInWithEmail: jest.fn(),
        _signIn: jest.fn(),
        setUser: jest.fn(),
        signInWithOAuth: jest.fn(),
        signOut: mockSignOut,
      }));
      render(<DeleteAccountButton user={TEST_USER} />);
      const btn = screen.getByRole('button', {
        name: 'main.profile.button.delete-account',
      });
      await userEvent.click(btn);
    });

    test('on Success', async () => {
      simpleMockHttp({
        method: 'delete',
        path: '/api/admin/users/{id}',
        params: { id: TEST_USER.id },
      });

      const deleteBtn = screen.getByRole('button', {
        name: 'button.delete',
      });
      await userEvent.click(deleteBtn);
      await waitFor(() => expect(mockSignOut).toHaveBeenCalled());
    });
    test('on Error', async () => {
      simpleMockHttp({
        method: 'delete',
        path: '/api/admin/users/{id}',
        params: { id: TEST_USER.id },
        status: 500,
      });

      const deleteBtn = screen.getByRole('button', {
        name: 'button.delete',
      });
      await userEvent.click(deleteBtn);

      await waitFor(() =>
        expect(screen.getByText(new RegExp('error', 'i'))).toBeInTheDocument(),
      );
      await waitFor(() => expect(mockSignOut).not.toHaveBeenCalled());
    });
  });
});
