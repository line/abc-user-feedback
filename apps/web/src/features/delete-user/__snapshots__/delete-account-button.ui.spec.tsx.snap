// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteAccountButton match snapshot 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="btn btn-md btn-secondary text-red-primary min-w-[120px]"
    data-state="closed"
    type="button"
  >
    main.profile.button.delete-account
  </button>
  <div
    style="position: fixed; z-index: 9999; top: 16px; left: 16px; right: 16px; bottom: 16px; pointer-events: none;"
  />
</div>
`;
